source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.h

```diff
@@ -0,0 +1,121 @@
+  // entries where the version ID does not match.
+  uint32_t cacheVersionId() const {
+    return fromEndianness<ByteOrder::BigEndian>(contents_.cache_version_id_);
+  }
+
+  // Size of the serialized proto message capturing headers and metadata.
+  size_t headerSize() const { return fromEndianness<ByteOrder::BigEndian>(contents_.header_size_); }
+
+  // Size of the serialized proto message capturing trailers.
+  size_t trailerSize() const {
+    return fromEndianness<ByteOrder::BigEndian>(contents_.trailer_size_);
+  }
+
+  // Size of the http body of the cache entry.
+  size_t bodySize() const { return fromEndianness<ByteOrder::BigEndian>(contents_.body_size_); }
+
+  // Sets the size of the serialized headers, key and metadata.
+  void setHeadersSize(size_t sz) {
+    contents_.header_size_ = toEndianness<ByteOrder::BigEndian>(static_cast<uint32_t>(sz));
+  }
+
+  // Sets the size of the serialized trailers.
+  void setTrailersSize(size_t sz) {
+    contents_.trailer_size_ = toEndianness<ByteOrder::BigEndian>(static_cast<uint32_t>(sz));
+  }
+
+  // Sets the size of the serialized body.
+  void setBodySize(size_t sz) {
+    contents_.body_size_ = toEndianness<ByteOrder::BigEndian>(static_cast<uint64_t>(sz));
+  }
+
+  // Returns the offset from the start of the file to the start of the
+  // serialized headers proto.
+  static size_t offsetToHeaders() { return size(); }
+
+  // Returns the offset from the start of the file to the start of the body data.
+  size_t offsetToBody() const { return offsetToHeaders() + headerSize(); }
+
+  // Returns the offset from the start of the file to the start of the
+  // serialized trailers proto.
+  size_t offsetToTrailers() const { return offsetToBody() + bodySize(); }
+
+  // Returns a string_view of the serialized fixed header chunk.
+  absl::string_view stringView() const { return {contents_.as_str_, size()}; }
+
+  // Returns true if the fileId and cacheVersionId match the compile-time constants from
+  // cache_file_fixed_block.cc
+  bool isValid() const;
+
+private:
+  // sets the fileId, a fixed value used to identify that this is a cache file.
+  // This is private so only the unit test can call it, as the file ID should be
+  // a compile-time constant in ordinary use.
+  void setFileId(uint32_t id) { contents_.file_id_ = toEndianness<ByteOrder::BigEndian>(id); }
+
+  // sets the cacheVersionId, a value that should be consistent between versions
+  // of the file cache implementation.
+  // This is private so only the unit test can call it, as the version ID should be
+  // a compile-time constant in ordinary use.
+  void setCacheVersionId(uint32_t id) {
+    contents_.cache_version_id_ = toEndianness<ByteOrder::BigEndian>(id);
+  }
+
+  union {
+    struct {
+      uint32_t file_id_;
+      uint32_t cache_version_id_;
+      uint32_t header_size_;
+      uint32_t trailer_size_;
+      uint64_t body_size_;
+    };
+    char as_str_[1];
```

Comment:
suggest renaming to 'data_'. I see what you are trying to do here but think calling it 'data 'is more helpful than "as string".

Reply:
Disagree: data_ doesn't really mean anything; all the other values in the union are also data. The purpose of the union with as_str_ is to provide a string-like interface. Calling it data wouldn't even hint at what its purpose is.

Reply:
I was trying to stay analogous to std::string::data() and absl::string_view::data(), both of which can only be used in the context of a size(), and do not null-terminate, which is why I might expect for something that claims to be a "string", but provides only a char*.

The pattern you give looks a lot more like a structure with variable-length array at the end, which at first is what I thought you were doing until I read more closely.

This whole block of code is correct -- not disputing that -- but it took me a while to understand why you had done it this way and why it's safe. This is why my higher level thought is to drop all this in favor of a s protobuf with 4 integer members, and use protobuf utils to serialize/deserialize. When people trying to ensure that Envoy is secure see code that hacks bytes like this it raises a lot of questions about security and correctness that are easier to reason about if we work at a higher level. I will not be the last one to read this code and need to think deeply about it.

As I said there are times when you need to go low-level but I think our default position is that higher level constructs are used unless we document why we are not doing that, in the place where we have the low-level byte-hacking code.

Reply:
Explanation for why not protobuf is now in the class comment.
The difference between this and those string classes is that in the string classes "data" is the sole underlying representation of all the data the class wraps. In this case, the struct of variables is *also* the underlying representation of all the data, so the word data could mean that, or it could mean the raw data (as it would), which makes its lack of meaning a problem. Renamed to raw_, to clarify that it's not a str, and gave it an explicit size to clarify that it's not variable-length.